includes/Core/Nonces/Nonces.php

```diff
@@ -0,0 +1,138 @@
+<?php
+/**
+ * Class Google\Site_Kit\Core\Nonces\Nonces
+ *
+ * @package   Google\Site_Kit\Core\Nonces
+ * @copyright 2021 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\Nonces;
+
+use Google\Site_Kit\Context;
+use Google\Site_Kit\Core\Authentication\Authentication;
+use Google\Site_Kit\Core\Permissions\Permissions;
+use Google\Site_Kit\Core\REST_API\REST_Route;
+use Google\Site_Kit\Core\REST_API\REST_Routes;
+use Google\Site_Kit\Core\Util\Feature_Flags;
+use WP_REST_Response;
+use WP_REST_Server;
+
+/**
+ * Class managing plugin permissions.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+final class Nonces {
+	/*
+	 * Nonce actions
+	 *
+	 * @since n.e.x.t
+	 */
+	const NONCE_UPDATES = 'updates';
+
+	/**
+	 * Plugin context.
+	 *
+	 * @since n.e.x.t
+	 * @var Context
+	 */
+	private $context;
+	/**
+	 * Array of nonce actions.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	private $nonce_actions;
+
+	/**
+	 * Constructor.
+	 *
+	 * Sets up the capability mappings.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Context $context Plugin context.
+	 */
+	public function __construct( Context $context ) {
+		$this->context = $context;
+
+		$this->nonce_actions = array(
+			self::NONCE_UPDATES,
+			'testnonce',
+		);
+	}
+
+	/**
+	 * Registers functionality through WordPress hooks.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function register() {
+		add_filter(
+			'googlesitekit_rest_routes',
+			function( $routes ) {
+				return array_merge( $routes, $this->get_rest_routes() );
+			}
+		);
+
+		add_filter(
+			'googlesitekit_apifetch_preload_paths',
+			function ( $paths ) {
+				return array_merge(
+					$paths,
+					array(
+						'/' . REST_Routes::REST_ROOT . '/core/user/data/nonces',
+					)
+				);
+			}
+		);
+	}
+
+	/**
+	 * Generate nonces for the current user.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array List of nonces.
+	 */
+	public function get_nonces() {
+		$nonces = array();
+
+		foreach ( $this->nonce_actions as $nonce_action ) {
+			$nonces[ $nonce_action ] = wp_create_nonce( $nonce_action );
+		}
+
+		return $nonces;
+	}
```

Comment:
There's already a URL we can use to get nonces for the user; it's used in our own API shim library: https://github.com/google/site-kit-wp/blob/78a3d18dde7794cc0e3cfccad4bb15742b4d236f/assets/js/api-fetch-shim.js#L29-L33

I know we aren't using that library, possibly because of `createRootURLMiddleware`; I wonder if we could create another API fetch instance without that middleware that would allow us to request `/admin-ajax.php`… but at the version least could we use the values from `global._googlesitekitAPIFetchData` to get a nonce-creation URL rather than creating our own route to create nonces?

Reply:
I don't think it is possible to creating a nonce middleware for `/admin-ajax.php`. The REST API requires the nonces to be sent as a request header. This makes it easy to create a middleware.
But ajax actions don't have such a convention about implementing nonces.

The `toggle-auto-updates` ajax action we use here, requires the nonce to be send in the body as url-form-encoded. Therefore the nonce is added together with the other request data.
https://github.com/google/site-kit-wp/blob/0d954750d4df376168e146af34fe1a469b6a299d/assets/js/googlesitekit/datastore/site/enable-auto-update.js#L39-L41

Most ajax actions require the nonce being sent like this, although this could be different for other ajax requests as WordPress does not enforce a specific way of sending nonces via ajax.

Reply:
Ah, okay, that makes sense. Thanks! 👍🏻